pkg/endpoint/endpoint.go

```diff
@@ -345,51 +346,18 @@ type Endpoint struct {
 	// proxy redirects to remove later during the build
 	proxiesToRemove map[string]bool
 
-	// Pending completions, protected by the BuildMutex!
-	ProxyCompletions *Completions
-}
-
-// Completions maintains the state needed to register and wait for asynchronous completions.
-type Completions struct {
-	wg     sync.WaitGroup
-	lock   lock.Mutex
-	errors int
-	done   bool
-}
-
-// Completed is called when an asynchronous event is completed.
-// May be called from any goroutine without holding any locks
-func (c *Completions) Completed(success bool) {
-	log.Debug("completions.Completed: ", success)
-	// 'done' is used to catch late callbacks (after Wait()), non-locked access on purpose.
-	if c.done {
-		log.Fatal("Completed called after Wait!")
-	}
-	if !success {
-		c.lock.Lock()
-		c.errors++
-		c.lock.Unlock()
-	}
-	c.wg.Done()
-}
-
-// AddCompletion increases the count of 'Completed()' callbacks we need by 1.
-// Called with BuildMutex held
-func (e *Endpoint) AddCompletion() (policy.Completion, time.Duration) {
-	e.ProxyCompletions.wg.Add(1)
-	return e.ProxyCompletions, time.Duration(10) * time.Second
+	// ProxyWaitGroup waits for pending proxy changes to complete.
```

Comment:
Is this also protected by the BuildMutex? If so, I'd like to keep that comment.

Reply:
It doesn't need to.

Reply:
OK, thanks.

Reply:
I haven't removed any locking using BuildMutex. Whatever locking was done is still there.
Sorry I might have misunderstood @ianvernon's original question.

Reply:
This comment thread is due to your removal of wording "protected by the BuildMutex!" from the comment. It needs to be brought back to help avoid races in future.